doc/developer/design/20220511_replicated_compute_introspection.md

```diff
@@ -0,0 +1,360 @@
+view contains the union of all replica-specific introspection tables for the
+respective log variant, tagged with cluster name and replica name:
+
+```sql
+CREATE VIEW <mz-log-variant>_persist AS
+  (
+    SELECT 'c1' AS cluster, 'r1' AS replica, *
+    FROM <mz-log-variant>_1
+  ) UNION (
+    SELECT 'c1' AS cluster, 'r2' AS replica, *
+    FROM <mz-log-variant>_2
+  ) UNION (
+    SELECT 'c2' AS cluster, 'r1' AS replica, *
+    FROM <mz-log-variant>_3
+  ) ...
+```
+
+Note that the `_persist` (or some other) suffix to the view name is required
+because the old arrangement-based introspection tables already occupy the plain
+`<mz-log-variant>` names. Once we remove the arrangement-based introspection
+tables, we can also drop the suffix.
+
+When a new replica is created, the coordinator adds its introspection tables to
+the respective introspection views.
+
+When a replica is dropped, the coordinator removes its introspection tables from
+the respective introspection views.
+
+### Feature Flag
+
+As mentioned previously, we don't want to enable logging to persist by default
+as long as persist does not support compaction, as otherwise we would risk
+unbounded growth of the introspection collections. Therefore, logging to
+persist is disabled by default.
+
+We add a feature flag `--replicated-compute-introspection` to both
+`materialized` and `computed`, to opt into logging introspection data to
+persist. `materialized` passes the flag (if set) on to `computed`.
+
+If the flag is not set, the coordinator does not create the introspection tables
+and views described above. It may still generate shard IDs and pass them to the
+compute controller, as this is invisible to users.
+
+If the flag is not set, replicas don't sink introspection data to persist.
+
+## Future Improvements
+
+Once persist supports compaction, the risk of unbounded storage usage is
+mitigated and we can enable replicated compute introspection by default. At that
+point, we can also drop the old arrangement-based introspection tables.
+
+Once a STORAGE client that abstracts persist is available, replicas should sink
+introspection data using this client instead of invoking `persist` directly. We
+expect the STORAGE client to (eventually) support sharded collections, which
+would allow us to distribute the work of writing and reading introspection data
+between all available compute workers.
+
+Once the STORAGE client supports renditions, we can make use of that feature to
+clear introspection collections without potentially running into "zombie
+replica" issues. Instead of writing retraction updates to the same persist
+shards, replicas can, upon restart, create new renditions of their introspection
+collections and continue writing to these. All concurrent readers will
+transparently switch to the new rendition. If there are previous incarnations of
+the replica still running, they will continue writing to the old rendition,
+without causing any harm.
```

Comment:
I'm not sure if this is correct. Won't the previous incarnations also get notified about the rendition change via the metadata collection and switch to the new rendition as well?

Reply:
This does seem like a slightly different use for renditions than I've heard, but I'm curious what the storage team thinks. Perhaps even in a storage collection world, we might want to give each replica its own storage collection. That would maybe clarify some of the semantics around how long the old replica's debugging information sticks around for example (the old replica's storage collection would have its own since). It might even save some complexity and time from issuing retractions on replica startup (the new replica would start with an empty collection).

Reply:
Each replica should definitely have its own collection! I assume what you mean is that each *version*/*incarnation* of a replica should have its own collection (i.e., when a replica is restarted, it sinks to a new collection instead of reusing the existing one)?

There are two potential issues with that approach:
* Cleanup: It is possible that the previous incarnation of the replica wasn't able to shut down cleanly and didn't have opportunity to delete its collection. How do we make sure that superseded introspection collections are always cleaned up?
* Concurrent readers: There might be long-running dataflows that read from introspection collections (imagine having a cluster that continuously monitors the state of the system). If a replica is restarted and creates new introspection collections, those dataflows will continue reading from the old collections. How do we make sure they start reading from the new collections at the correct time?